File: 203.remove-linked-list-elements.c
/*
 * @lc app=leetcode id=203 lang=c
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* removeElements(struct ListNode* head, int val)
{
    if (!head) {
        return NULL;
    }
    struct ListNode *ret = head;
    struct ListNode *prev = NULL;
    struct ListNode *node = NULL;
    while (head) {
        if (head->val == val) {
            node = head;
            if (prev) {
                prev->next = head->next;
            }
            else {
                ret = head->next;
            }
            head = head->next;
            free(node);
            continue;
        }
        prev = head;
        head = head->next;
    }
    return ret;
}
// @lc code=end

File: 203.remove-linked-list-elements.go
/*
 * @lc app=leetcode id=203 lang=golang
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeElements(head *ListNode, val int) *ListNode {
	// 还是快慢指针
	dummy := &ListNode{Next: head}

	slow := dummy
	fast := head

	for fast != nil {
		// 遇到指定元素，慢指针的下一个等于快指针的下一个
		// 跳过指定元素的列表
		if fast.Val == val {
			slow.Next = fast.Next
		} else {
			// 否则慢指针=快指针
			slow = fast
		}
		fast = fast.Next
	}
	return dummy.Next
}

// @lc code=end
